### PR TITLE
DecisionTree Refactor

### PR DESCRIPTION
**Pull request by varunagrawal**
Refactored `DecisionTree` to:

1. Make a distinction between leaves and assignments. A tree always has a full number of assignments (which each gives us a value) but can have a smaller number of leaves (the data structure) due to pruning.
2. Updated the documentation accordingly.
3. Renamed `DT_NO_PRUNING` to `GTSAM_DT_NO_PRUNING` in the event we decide to expose it via CMake.

### CODE REVIEW

**Comment by varunagrawal** [comment]
@dellaert hope this PR is up to your expectations.

**Comment by varunagrawal** [comment]
@dellaert gentle reminder to take a look at this when you can :)

File: gtsam/discrete/DecisionTree-inl.h [hunk 1]
```diff
@@ -678,7 +680,14 @@ namespace gtsam {
   }
 
   /****************************************************************************/
-  // Functor performing depth-first visit without Assignment<L> argument.
+  /**
+   * Functor performing depth-first visit without Assignment<L> argument.
+   *
+   * NOTE: We differentiate between leaves and assignments. Concretely, a 3
```
Comment (dellaert) [inline gtsam/discrete/DecisionTree-inl.h:686]: So, what does this Functor do? Leaves, right? And should it not now pass the number of assignments captured by the leaf?

File: gtsam/discrete/DecisionTree-inl.h [hunk 2]
```diff
@@ -678,7 +680,14 @@ namespace gtsam {
   }
 
   /****************************************************************************/
-  // Functor performing depth-first visit without Assignment<L> argument.
+  /**
+   * Functor performing depth-first visit without Assignment<L> argument.
```
Comment (dellaert) [inline gtsam/discrete/DecisionTree-inl.h:684]: What *is* the argument then?

File: gtsam/discrete/DecisionTree-inl.h [hunk 3]
```diff
@@ -763,12 +775,14 @@ namespace gtsam {
   }
 
   /****************************************************************************/
-  // labels is just done with a visit
+  // Get (partial) labels by performing a visit.
```
Comment (dellaert) [inline gtsam/discrete/DecisionTree-inl.h:778]: Explain more. Maybe with example.

File: gtsam/discrete/DecisionTree.h [hunk 4]
```diff
@@ -234,11 +233,13 @@ namespace gtsam {
 
     /**
      * @brief Visit all leaves in depth-first fashion.
-     * 
-     * @param f side-effect taking a value.
-     * 
-     * @note Due to pruning, leaves might not exhaust choices.
-     * 
+     *
```
Comment (dellaert) [inline gtsam/discrete/DecisionTree.h:236]: I think these functions passed in  should now also take an argument for the number of assignments captured in a particular leaf.

File: gtsam/discrete/DecisionTree-inl.h (hunk 1 above)
Comment (varunagrawal) [reply gtsam/discrete/DecisionTree-inl.h:686]: The `VisitLeaf` functor passes the `Leaf` object which has the number of assignments recorded in it.

File: gtsam/discrete/DecisionTree.h (hunk 4 above)
Comment (varunagrawal) [reply gtsam/discrete/DecisionTree.h:236]: I added `visitLeaf` as an alternative to that. Should I remove that and update the rest to do this (will be API breaking)?

**Review by dellaert — APPROVED** [review]
Approved, and I will merge...

File: gtsam/discrete/DecisionTree.h (hunk 4 above)
Comment (dellaert) [reply gtsam/discrete/DecisionTree.h:236]: Shoot, I merged too soon! I just noticed this new visitLeaf, which in this PR seems identical??

Comment (dellaert) [reply gtsam/discrete/DecisionTree.h:236]: And the examples are wrong I think... We can talk about it in our meeting.

Comment (varunagrawal) [reply gtsam/discrete/DecisionTree.h:236]: Okay! Though I don't want to spend more than 5 minutes on this since it is secondary to the workshop paper.

**Pull request merged**